### ISSUE TITLE
Security account for new instance on GCP

### ISSUE THREAD

**Issue by imarsman** [nanovms/ops#1151]
Hello. The gcloud cli will apparently attach the default security account to an instance created using that method. However, it is noted that a security account is not attached when creating an instance using some other means (not really specified). I have found that when using OPS to create an instance from an image there is no security account attached. Also, the image is created in a started state. I could be missing something. However, to work around this issue I have added calls to stop the image, attach the account and desired scope(s), then restart. This seems to work. As I have said, I could be wrong about what is happening, but if not, would it be possible to create an instance in a stopped state to avoid having to stop it before adding the security account? I really am enjoying OPS and how well it is able to create images and instances and have found calls to the gcloud cli to integrate well with ops to add in actions more appropriately handled there.

**Comment by imarsman** [comment]
Stopping the new instance, attaching security account and scope, then restarting works. I am able to use cloud logs, which has been my initial test.

**Comment by rrs45** [comment]
Just curious, are you able to see stdout logs from the app in nanos
instance, in cloud logs?

**Comment by imarsman** [comment]
Yes, with fmt.Println at least. That is actually helpful. I think I am seeing a race condition between the instance starting and trying to use the GCE api before the system itself is ready to assess authorization. I am changing things to wait for a few seconds for now then manually call the logging initialization then try to do cloud logging. There is likely a much more sophisticated way to do this.

**Comment by imarsman** [comment]
So the Google Cloud API I am using at the moment is cloud logging. In my main method I added a 30 second sleep then called a function I'd written that used to be done in init() to start up cloud logging. The logic was to give GCP a chance to really get ready to interact with the instance. Given that unikernels seem to start up a lot faster than things like Linux VMs perhaps their preparation to interact with a unikernel VM is less synchronous than it should be. For now I'll leave the lame sleep and look into ways to "ping" to see if cloud logging can be done then wait a few seconds for a maximum number of tries.

**Comment by imarsman** [comment]
I've added a function to wait until the metadata service for the instance is available. It doesn't really seem to have to wait long to get at the service and in this code the client timeout might not expire so I'm not sure how long it waits, but here is the code I have now:

// TestPermissions test to see if API can be used
func TestPermissions() error {
	c := make(chan int)

	var err error
	go func() {
		for i := 0; i < 10; i++ {
			client := metadata.NewClient(http.DefaultClient)
			_, err := client.ProjectID()
			if err != nil {
				fmt.Println("Waiting")
				time.Sleep(2 * time.Second)
			} else {
				fmt.Println("Finished wait.")
				c <- 1
				return
			}
		}
		c <- 1
	}()

	<-c
	return err
}

It never prints "Waiting", so I am assuming the client acquisition timeout is not exceeded. Evidently this needs more exploration, but so far I seem to be reliably able to use this waiting technique then use the cloud log.

**Comment by imarsman** [comment]
Hello! I apologize for my nomenclature error. Yes, I do mean service account. My previous work has mostly been with AWS and I'm still acclimatizing.

Ops does of course require permissions to do things like create images and instances. That is working just fine, though deleting an instance has an interaction with the underlying image, but that is separate and perhaps my error.

I have noticed that when creating an instance, the new instance does not have a service account attached and the instance is in a started state. To remedy this I have added steps to stop the instance, modify the security account attachment, then restart it. This has allowed my code to do cloud logging and now to get the instance's IP. I want to move on to getting a list of public IPs for a Managed Instance Group.

I will re-investigate to ensure that the service account is not attached on initial instance creation. If the default is to have an instance be running on creation I can live with that. I'll test by removing my stop, attach, start steps and examining fmt.Println statements I use to write to the syslog.

I really love Ops and unikernels in general. I hope my questions and notes are useful for those doing work on Ops and nanovm. Thank-you for everything.

**Comment by eyberg** [comment]
I see - attaching the svc account is not going to install a cloud agent on the instance, which is perhaps what you are expecting?

as for getting the instance ip - both priv/pub ips should be present immediately regardless showing up in 'ops instance list'
